### ISSUE TITLE
Typescript: $assertPopulated does not preserve the correct type of the document

### ISSUE THREAD

**Issue by ajwootto** [Automattic/mongoose#12234]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

6.5.1

### Node.js version

16.13.2

### MongoDB server version

5.x

### Description

The recently-added $assertPopulated helper currently does not preserve the "document" type of a document instance, and instead just turns it into the underlying interface type:

```typescript
import { Schema, model, connect, Types } from 'mongoose'

interface NestedChild {
    name: string
    _id: Types.ObjectId
}
const nestedChildSchema: Schema = new Schema({ name: String })

interface Parent {
    nestedChild: Types.ObjectId
    name?: string
}

const ParentModel = model<Parent>('Parent', new Schema({
    nestedChild: { type: Schema.Types.ObjectId, ref: 'NestedChild' },
    name: String
}))

const NestedChildModel = model<NestedChild>('NestedChild', nestedChildSchema)

async function run() {
    await connect('mongodb://localhost:27017/')
    const parent = (new ParentModel({
        nestedChild: new NestedChildModel({ name: 'test' }),
        name: 'Parent'
    })).$assertPopulated<{nestedChild: NestedChild}>('nestedChild')

    // works, assertPopulated makes sure `parent` has a nestedChild field correctly typed
    console.log(parent.nestedChild.name)

   // doesn't work, parent is no longer a Document 
    await parent.save()
}

run()
```

From what I can tell, changing the definition from this:
```typescript
$assertPopulated<Paths = {}>(paths: string | string[]): Omit<DocType, keyof Paths> & Paths;
```

to this:
```
$assertPopulated<Paths = {}>(paths: string | string[]): Omit<this, keyof Paths> & Paths;
```

would fix the problem.

### Steps to Reproduce

See code snippet

### Expected Behavior

I would expect that when working with a document instance, $assertPopulated should preserve the document type so that it's possible to save it etc.

**Comment by vkarpov15** [comment]
We'll look into this, however, we switched `this` -> `DocType` for performance reasons. Using `this` caused some serious slowdowns in tsc.

**Comment by vkarpov15** [comment]
I switched `this` back because I confirmed that it doesn't cause a performance regression anymore. I think the performance issues with `this` were exacerbated by some of the issues we fixed in #10349.

Fix will be in 6.5.4 :+1:
